### ISSUE TITLE
vectorize bootstrap CI calculations

### ISSUE THREAD

**Issue by jdherman** [jdherman/SALib#47]
for other methods besides Morris. Also consider reporting the standard error in the results instead of the CI half-width, because this has been confusing for people.

**Comment by willu47** [comment]
Could we put this code in the `utils` module and add it to the imports in each analyze module?

**Comment by jdherman** [comment]
Sure we can think about it ... but IIRC the calculations for each method work a little differently, so we might not be able to generalize it into a function.

There is probably a way. This is a good idea, I'll think about it.

**Comment by willu47** [comment]
There seem to be many other SciPy dependencies too e.g.

`from scipy.stats import norm, gaussian_kde, rankdata` in `delta.py` and 
`from scipy.spatial.distance import cdist`, `from scipy.misc import comb as nchoosek` in `morris_util.py`

**Comment by willu47** [comment]
I note that sobol.py has some redundant code too:

```python
def first_order_confidence(A, AB, B, num_resamples, conf_level):
    s = np.empty(num_resamples)
    for i in range(num_resamples):
        r = np.random.randint(len(A), size=len(A))
        s[i] = first_order(A[r], AB[r], B[r])

    return norm.ppf(0.5 + conf_level / 2) * s.std(ddof=1)

def total_order_confidence(A, AB, B, num_resamples, conf_level):
    s = np.empty(num_resamples)
    for i in range(num_resamples):
        r = np.random.randint(len(A), size=len(A))
        s[i] = total_order(A[r], AB[r], B[r])

    return norm.ppf(0.5 + conf_level / 2) * s.std(ddof=1)

def second_order_confidence(A, ABj, ABk, BAj, B, num_resamples, conf_level):
    s = np.empty(num_resamples)
    for i in range(num_resamples):
        r = np.random.randint(len(A), size=len(A))
        s[i] = second_order(A[r], ABj[r], ABk[r], BAj[r], B[r])

    return norm.ppf(0.5 + conf_level / 2) * s.std(ddof=1)
```

The only difference between these is the call to the first_order, total_order and second_order procedures.

So create a wrapper function which sets up the data, and then call the vectorised function on this data?

**Comment by willu47** [comment]
Comparison between old and new morris:

```python

def compute_mu_star_confidence(ee, num_trajectories, num_resamples, conf_level):

    ee_resampled = np.empty([num_trajectories])
    mu_star_resampled = np.empty([num_resamples])

    if conf_level < 0 or conf_level > 1:
        raise ValueError("Confidence level must be between 0-1.")

    for i in range(num_resamples):
        for j in range(num_trajectories):

            index = np.random.randint(0, num_trajectories)
            ee_resampled[j] = ee[index]

        mu_star_resampled[i] = np.average(np.abs(ee_resampled))

    return norm.ppf(0.5 + conf_level / 2) * mu_star_resampled.std(ddof=1)

def compute_mu_star_confidence(ee, num_trajectories, num_resamples, conf_level):
    '''
    Uses bootstrapping where the elementary effects are resampled with replacement
    to produce a histogram of resampled mu_star metrics.
    This resample is used to produce a confidence interval.
    '''
    ee_resampled = np.empty([num_trajectories])
    mu_star_resampled = np.empty([num_resamples])

    if conf_level < 0 or conf_level > 1:
        raise ValueError("Confidence level must be between 0-1.")

    resample_index = np.floor(np.random.rand(num_resamples * num_trajectories)*len(ee)).astype(int).reshape(num_resamples, num_trajectories)
    ee_resampled = ee[resample_index]
    # Compute average of the absolute values over each of the resamples
    mu_star_resampled = np.average(np.abs(ee_resampled), axis=1)

    return norm.ppf(0.5 + conf_level / 2) * mu_star_resampled.std(ddof=1)

```

**Comment by jdherman** [comment]
Right, duh, can't get rid of the Scipy dependency. Thanks for pointing that out. I guess we can keep the conf_level too instead of returning the standard error.

Yea those sobol calculations can be cleaned up A LOT. That's on me. Thanks for pulling out the Morris example, those loops I had in there are cringe-worthy :)

**Comment by willu47** [comment]
He he.  I'm sorry, looking back at earlier work is always a bit painful :-1:

**Comment by jdherman** [comment]
Alright, I came up with a way to modify the index calculations to work on either vectors or matrices:
```python
def first_order(A, AB, B):
    # First order estimator following Saltelli et al. 2010 CPC, normalized by
    # sample variance
    return np.mean(B * (AB - A), axis=0) / np.var(np.r_[A, B], axis=0)

def total_order(A, AB, B):
    # Total order estimator following Saltelli et al. 2010 CPC, normalized by
    # sample variance
    return 0.5 * np.mean((A - AB) ** 2, axis=0) / np.var(np.r_[A, B], axis=0)
```

So if you call them with regular vectors you get the value itself. But with a resample index matrix it can be used to compute the CI half-width:
```python
r = np.random.randint(N, size=(N, num_resamples))
Z = norm.ppf(0.5 + conf_level / 2)

    for j in range(D):
        S['S1'][j] = first_order(A, AB[:, j], B)
        S['S1_conf'][j] = Z * first_order(A[r], AB[r,j], B[r]).std(ddof=1)
        # ...
```

Unfortunately, and bizarrely, this actually slows down the Sobol calculations by about a factor of 1.5. It is much cleaner though because it removes all those functions you mentioned before. To deal with this I suggest changing the default number of resamples from 1000 to 100. 

Working on a pull request for this and a couple other things.

**Comment by jdherman** [comment]
Addressed by #54 ... the code is much cleaner now, but somehow slower. We'll keep an eye on it.

The first and total order functions now operate on either vectors or matrices, for example:
```python
def first_order(A, AB, B):
    # First order estimator following Saltelli et al. 2010 CPC, normalized by
    # sample variance
    return np.mean(B * (AB - A), axis=0) / np.var(np.r_[A, B], axis=0)
```

To do the bootstrap calculations, pass in `(A[r], AB[r,j], B[r])` where `r = np.random.randint(N, size=(N, num_resamples))`. I think the size of these matrices being passed around is what's causing the slowdown. To offset this for now I reduced the default number of bootstrap resamples from 1000 to 100, which should be plenty.
